testing/decorators.py

```diff
@@ -0,0 +1,162 @@
+import os
+from inspect import isclass
+from copy import deepcopy
+
+import pytest
+import shutil
+import textwrap
+
+from .utils import deep_update
+
+__all__ = ['default_confoverrides', 'confoverrides', 'no_run']
+
+
+def _apply_confoverride_to_class(cls, config, priority):
+    """
+    Apply a configuration override ``config`` to class ``cls`` with a given priority.
+
+    We need the priority trick as the default configuration is applied before a possible
+    class-wide decorator, which should supersede the default configuration.  We use
+    ``pytest.mark.exhale`` as a store of ``kwargs`` to apply to ``pytest.mark.sphinx``,
+    and we use the priority to combine these kwargs with respect to priorities.
+    """
+
+    # look for test methods in the class
+    for name, meth in cls.__dict__.items():
+        if not callable(meth) or not name.startswith('test_'):
+            continue
+
+        # meth is a test method, let's go and mark it!
+
+        # create marker kwargs
+        kwargs = dict(confoverrides=deepcopy(config))
+
+        # apply the exhale marker to the method, so that priority is saved
+        meth = pytest.mark.exhale(priority, **kwargs)(meth)
+
+        # create a list of exhale markers kwargs
+        markers_kwargs = []
+        for m in meth.exhale._marks:
+            markers_kwargs.append((m.args[0], m.kwargs))
+        # sort that list according to priority
+        markers_kwargs.sort(key=lambda m: m[0])
+
+        # now we can generate the sphinx fixture kwargs by combining the above list of
+        # kwargs depending on priority
+        sphinx_kwargs = {'testroot': cls.testroot}
+        for __, kw in markers_kwargs:
+            deep_update(sphinx_kwargs, kw)
+
+        ################################################################################
+        # Automatically create docs/{conf.py,index.rst} for the test project.          #
+        ################################################################################
+        def _with_docs_dir(self):
+            # Create the test project's 'docs' dir with a conf.py and index.rst.
+            if os.path.isdir(self.testroot):
+                shutil.rmtree(self.testroot)
+            os.makedirs(self.testroot)
+            # Sphinx demands a `conf.py` is present
+            with open(os.path.join(self.testroot, "conf.py"), "w") as conf_py:
+                conf_py.write(textwrap.dedent('''\
+                    # -*- coding: utf-8 -*-
+                    extensions = ["breathe", "exhale"]
+                    master_doc = "index.rst"
+                '''))
+
+            # If a given test case needs to run app.build(), make sure index.rst
+            # is available as well
+            cls_exhale_args = sphinx_kwargs["confoverrides"]["exhale_args"]
```

Comment:
@tkhyn Ok so this is where my understanding breaks.

1. The goal is automatically creating `testing/projects/{test_project}/docs/{conf.py,index.rst}`.
    - Currently this is done with `scope=class`, which is a problem when `containmentFolder` or `rootFileName` are changed in `conf.py`.
    - Not a huge deal (I will document it more explicitly, see `testing.base.ExhaleTestCase.copyThisProject`)?
2. I don't understand what's going on with the priority based `@pytest.mark.exhale` and the `sphinx_kwargs`.  I've read the marking docs like three times, but don't understand where this priority actually comes from.

Basically, what are your thoughts on this?  The reason for switching to project local `docs/conf.py` is because if you run the tests in parallel with a single `TEST_DOCS_DIR/conf.py` things can break.  Creating it at the function scope may not solve this either though.

Reply:
I'd say the way to change the sphinx root directory used by ``sphinx.testing.fixtures.app`` is to use ``pytest.mark.parametrize`` for the ``"app_params"`` fixture (see ``sphinx.testing.fixtures``). Some digging will be needed to know where and with which arguments to call it.